<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="generator" content="litedown 0.7">
<title>adverSCarial, generate and analyze the vulnerability of scRNA-seq</title>
<style type="text/css">
body {
  font-family: sans-serif;
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 1.5;
  print-color-adjust: exact;
  -webkit-print-color-adjust: exact;
}
body, .abstract, code, .footnotes, footer, #refs, .caption { font-size: .9em; }
li li { font-size: .95em; }
ul:has(li > input[type="checkbox"]) { list-style: none; padding-left: 1em; }
*, :before, :after { box-sizing: border-box; }
a { color: steelblue; }
pre, img { max-width: 100%; }
pre { white-space: pre-wrap; word-break: break-word; }
pre code { display: block; padding: 1em; overflow-x: auto; }
code { font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace; }
:not(pre, th) > code, code[class], div > .caption { background: #f8f8f8; }
pre > code:is(:not([class]), .language-plain, .language-none, .plain), .box, .figure, .table { background: inherit; border: 1px solid #eee; }
pre > code {
  &.message { border-color: #9eeaf9; }
  &.warning { background: #fff3cd; border-color: #fff3cd; }
  &.error { background: #f8d7da; border-color: #f8d7da; }
}
.fenced-chunk { border-left: 1px solid #666; }
.code-fence {
  opacity: .4;
  border: 1px dashed #666;
  border-left: 2px solid;
  &:hover { opacity: inherit; }
}
.box, .figure, .table, table { margin: 1em auto; }
div > .caption { padding: 1px 1em; }
.figure { p:has(img, svg), pre:has(svg) { text-align: center; } }
.flex-col { display: flex; justify-content: space-between; }
table {
  &:only-child:not(.table > *) { margin: auto; }
  th, td { padding: 5px; font-variant-numeric: tabular-nums; }
  thead, tfoot, tr:nth-child(even) { background: whitesmoke; }
  thead th { border-bottom: 1px solid #ddd; }
  &:not(.datatable-table) {
    border-top: 1px solid #666;
    border-bottom: 1px solid #666;
  }
}
blockquote {
  color: #666;
  margin: 0;
  padding: 1px 1em;
  border-left: .5em solid #eee;
}
hr, .footnotes::before { border: 1px dashed #ddd; }
.frontmatter { text-align: center; }
#TOC {
  a { text-decoration: none; }
  ul { list-style: none; padding-left: 1em; }
  & > ul { padding: 0; }
  ul ul { border-left: 1px solid lightsteelblue; }
}
.body h2 { border-bottom: 1px solid #666; }
.body .appendix, .appendix ~ h2 { border-bottom-style: dashed; }
.main-number::after { content: "."; }
span[class^="ref-number-"] { font-weight: bold; }
.ref-number-fig::after, .ref-number-tab::after { content: ":"; }
.cross-ref-chp::before { content: "Chapter "; }
.cross-ref-sec::before { content: "Section "; }
.cross-ref-fig::before, .ref-number-fig::before { content: "Figure "; }
.cross-ref-tab::before, .ref-number-tab::before { content: "Table "; }
.cross-ref-eqn::before, .MathJax_ref:has(mjx-mtext > mjx-c + mjx-c)::before { content: "Equation "; }
.abstract, #refs {
  &::before { display: block; margin: 1em auto; font-weight: bold; }
}
.abstract::before { content: "Abstract"; text-align: center; }
#refs::before { content: "Bibliography"; font-size: 1.5em; }
.ref-paren-open::before { content: "("; }
.ref-paren-close::after { content: ")"; }
.ref-semicolon::after { content: "; "; }
.ref-and::after { content: " and "; }
.ref-et-al::after { content: " et al."; font-style: italic; }
.footnote-ref a {
  &::before { content: "["; }
  &::after { content: "]"; }
}
section.footnotes {
  margin-top: 2em;
  &::before { content: ""; display: block; max-width: 20em; }
}
.fade {
  background: repeating-linear-gradient(135deg, white, white 30px, #ddd 32px, #ddd 32px);
  opacity: 0.6;
}

@media print {
  body { max-width: 100%; }
  tr, img { break-inside: avoid; }
}
@media only screen and (min-width: 992px) {
  body:not(.pagesjs) pre:has(.line-numbers):not(:hover) { white-space: pre; }
}
</style>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@xiee/utils@1.14.11/css/prism-xcode.min.css">
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/components/prism-core.min.js" defer></script>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/plugins/autoloader/prism-autoloader.min.js" defer></script>
</head>
<body>
<div class="frontmatter">
<div class="title"><h1>adverSCarial, generate and analyze the vulnerability of scRNA-seq</h1></div>
<div class="author"><h2>Ghislain FIEVET <a href="mailto:ghislain.fievet@gmail.com">ghislain.fievet@gmail.com</a></h2></div>
</div>
<div class="abstract"><blockquote>
</blockquote>
</div>
<div class="body">
<h1 id="chp:introduction">Introduction</h1>
<p><strong>adverSCarial is a package for generating and evaluating vulnerability to adversarial attacks on single-cell RNA sequencing classifiers.</strong>
Single cell analysis are on the way to be used in clinical routine. As a critical use, algorithms must address the challenge of ensuring reliability, one concern being the susceptibility to adversarial attacks.</p>
<p>Zhang, J., Wang, W., Huang, J., Wang, X., &amp; Zeng, Y. (2020). How far is single-cell sequencing from clinical application?. Clinical and translational medicine, 10(3), e117. <a href="https://doi.org/10.1002/ctm2.117">https://doi.org/10.1002/ctm2.117</a></p>
<p>de Hond, A.A.H., Leeuwenberg, A.M., Hooft, L. et al. Guidelines and quality criteria for artificial intelligence-based prediction models in healthcare: a scoping review. npj Digit. Med. 5, 2 (2022). <a href="https://doi.org/10.1038/s41746-021-00549-7">https://doi.org/10.1038/s41746-021-00549-7</a></p>
<p>The package is designed to generate and analyze the vulnerability of scRNA-seq classifiers to adversarial attacks. The package is versatile and provides a format for integrating
any type of classifier. It offers functions for studying and generating two types of attacks, single gene attack and max change attack.
The single gene attack involves making a small modification to the input to alter the classification. This is an issue when the change of classification is made on a gene that is not known to be biologicaly involved in the change of this cell type.
The max change attack involves making a large modification to the input without changing its classification. This is an issue when a significant percentage of genes, sometimes as high as 99%, can be modified with the classifier still predicting the same cell type.</p>
<h1 id="chp:jupyter-notebook-examples">Jupyter Notebook examples</h1>
<ul>
<li><a href="https://github.com/GhislainFievet/adverSCarial_article/blob/master/04_train_and_attack_scAnnotatR.ipynb">Train, format and CGD attack on a scAnnotatR classifier</a></li>
<li><a href="https://github.com/GhislainFievet/adverSCarial_article/blob/master/05_train_format_random_forest_classifier_scRF.ipynb">Train, format and max-change attack on a random forest classifier</a></li>
<li><a href="https://github.com/GhislainFievet/adverSCarial_article/blob/master/06_train_format_classifier_scMLP.ipynb">Train, format and single-gene attack on a multi layer perceptron classifier</a></li>
</ul>
<h1 id="chp:installation">Installation</h1>
<pre><code class="language-r">## Install BiocManager is necessary
if (!require(&quot;BiocManager&quot;)) {
    install.packages(&quot;BiocManager&quot;)
}
BiocManager::install('adverSCarial')
</code></pre>
<h1 id="chp:generate-an-adversarial-attack">Generate an adversarial attack</h1>
<p>There are three types of adversarial attacks: <code>single-gene attack</code> which modifies the expression of a gene in order to alter the classification, <code>max-change attack</code> which introduce the largest possible perturbations to the input while still keeping the same classification, and <code>Cluster Gradient Descent (CGD)</code> which modifies the input matrix gene by gene by following an approximated gradient until the cluster has switch its classification.</p>
<p>Load libraries</p>
<pre><code class="language-r">library(adverSCarial)
library(LoomExperiment)
</code></pre>
<p>Load a pbmc raw data</p>
<pre><code class="language-r">pbmcPath &lt;- system.file(&quot;extdata&quot;, &quot;pbmc_short.loom&quot;, package=&quot;adverSCarial&quot;)
lfile &lt;- import(pbmcPath, type=&quot;SingleCellLoomExperiment&quot;)
</code></pre>
<pre><code class="language-r">matPbmc &lt;- counts(lfile)
</code></pre>
<pre><code class="language-r">matPbmc[1:5, 1:5]
</code></pre>
<pre><code>## &lt;5 x 5&gt; DelayedMatrix object of type &quot;integer&quot;:
##      NEK11 IL5RA OR4C5 KHDRBS3 OPA3
## [1,]     0     0     0       0    0
## [2,]     0     0     0       0    0
## [3,]     0     0     0       0    0
## [4,]     0     0     0       0    0
## [5,]     0     0     0       0    0
</code></pre>
<p>and its cell type classification based on <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html">Seurat documentation</a></p>
<pre><code class="language-r">cellTypes &lt;- rowData(lfile)$cell_type
</code></pre>
<pre><code class="language-r">head(cellTypes)
</code></pre>
<pre><code>## [1] &quot;Naive CD4 T&quot;  &quot;Naive CD4 T&quot;  &quot;FCGR3A+ Mono&quot; &quot;Naive CD4 T&quot;  &quot;B&quot;           
## [6] &quot;Naive CD4 T&quot;
</code></pre>
<p>The package contains a marker based classifier working on the pbmc3k dataset: <code>MClassifier</code>.
We verify it is classifying properly according to the <code>cellTypes</code>.</p>
<pre><code class="language-r">for ( cell_type in unique(cellTypes)){
    resClassif &lt;- MClassifier(matPbmc, cellTypes, cell_type)
    print(paste0(&quot;Expected: &quot;, cell_type, &quot;, predicted: &quot;, resClassif[1]))
}
</code></pre>
<pre><code>## [1] &quot;Expected: Naive CD4 T, predicted: Naive CD4 T&quot;
## [1] &quot;Expected: FCGR3A+ Mono, predicted: FCGR3A+ Mono&quot;
## [1] &quot;Expected: B, predicted: B&quot;
## [1] &quot;Expected: Memory CD4 T, predicted: Memory CD4 T&quot;
## [1] &quot;Expected: CD14+ Mono, predicted: CD14+ Mono&quot;
## [1] &quot;Expected: CD8 T, predicted: CD8 T&quot;
## [1] &quot;Expected: UNDETERMINED, predicted: UNDETERMINED&quot;
## [1] &quot;Expected: NK, predicted: NK&quot;
## [1] &quot;Expected: Platelet, predicted: Platelet&quot;
## [1] &quot;Expected: DC, predicted: DC&quot;
</code></pre>
<p>We want to run attacks on the “DC” cluster but without modifying the genes used by human to make manual classification. We can find this list on <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html">Seurat documentation</a></p>
<pre><code class="language-r"># Known markers for each cell type
markers &lt;- c(&quot;IL7R&quot;, &quot;CCR7&quot;, &quot;CD14&quot;, &quot;LYZ&quot;, &quot;S100A4&quot;, &quot;MS4A1&quot;, &quot;CD8A&quot;, &quot;FCGR3A&quot;, &quot;MS4A7&quot;,
              &quot;GNLY&quot;, &quot;NKG7&quot;, &quot;FCER1A&quot;, &quot;CST3&quot;, &quot;PPBP&quot;)
</code></pre>
<h2 id="sec:single-gene-attack">Single-gene attack</h2>
<p>Modify the value of a gene inside a cell cluster in order to change its classification.</p>
<p>The function <code>advSingleGene</code> runs a dichotomic search of one gene attacks, given the cluster to attack and a type of modification. We suggest two main modifications: <code>perc1</code> replacing the value of the gene by its first percentile, and <code>perc99</code> replacing the value of the gene by its 99th percentile. The <code>adv_fct</code> argument allows users to choose custom modifications of the gene.</p>
<p>The <code>excl_genes</code> argument allows users to exclude certain genes from being modified. Here we exclude the genes usually used as markers of specific cell types, as defined previously.</p>
<p>Computation times can be lengthy, we use the <code>return_first_found</code> argument to return the result as soon as it is found.</p>
<p>We can specify the <code>change_type = &quot;not_na&quot;</code> argument to indicate that we want the attack to misclassify the cluster as an existing cell type, rather than as NA.</p>
<pre><code class="language-r">genesMinChange &lt;- advSingleGene(matPbmc, cellTypes, &quot;DC&quot;,
                        MClassifier, exclGenes = markers, advMethod = &quot;perc99&quot;,
                        returnFirstFound = TRUE, changeType = &quot;not_na&quot;,
                        firstDichot = 10)
</code></pre>
<pre><code class="language-r">genesMinChange
</code></pre>
<pre><code>## $PF4
## [1] &quot;Platelet&quot; &quot;1&quot;
</code></pre>
<p>The function found that modifying the single gene PF4 with the <code>perc99</code> modification on the cluster leads to a new classification, Platelet.</p>
<p>Let’s run this attack and verify if it is successful.</p>
<p>First we modify the <code>matPbmc</code> matrix on the target cluster.</p>
<pre><code class="language-r">matAdver &lt;- advModifications(matPbmc, names(genesMinChange@values)[1], cellTypes, &quot;DC&quot;)
</code></pre>
<p>Then classify the “DC” cluster with <code>MClassifier</code>.</p>
<pre><code class="language-r">resClassif &lt;- MClassifier(matAdver, cellTypes, &quot;DC&quot;)
</code></pre>
<pre><code class="language-r">resClassif
</code></pre>
<pre><code>## $prediction
## [1] &quot;Platelet&quot;
## 
## $odd
## [1] 1
## 
## $typePredictions
##              1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
## Naive CD4 T  0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## FCGR3A+ Mono 0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## B            0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Memory CD4 T 0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## CD14+ Mono   0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## CD8 T        0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## UNDETERMINED 0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## NK           0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Platelet     1 1 1 1 1 1 1 1 1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
## DC           0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
##              26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
## Naive CD4 T   0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## FCGR3A+ Mono  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## B             0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Memory CD4 T  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## CD14+ Mono    0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## CD8 T         0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## UNDETERMINED  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## NK            0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Platelet      1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
## DC            0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
##              48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69
## Naive CD4 T   0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## FCGR3A+ Mono  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## B             0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Memory CD4 T  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## CD14+ Mono    0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## CD8 T         0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## UNDETERMINED  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## NK            0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Platelet      1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
## DC            0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
##              70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91
## Naive CD4 T   0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## FCGR3A+ Mono  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## B             0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Memory CD4 T  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## CD14+ Mono    0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## CD8 T         0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## UNDETERMINED  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## NK            0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Platelet      1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
## DC            0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
##              92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109
## Naive CD4 T   0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## B             0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## CD14+ Mono    0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## CD8 T         0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## NK            0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## Platelet      1  1  1  1  1  1  1  1   1   1   1   1   1   1   1   1   1   1
## DC            0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
##              110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
## Naive CD4 T    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## B              0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD14+ Mono     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD8 T          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## NK             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Platelet       1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
## DC             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
##              126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141
## Naive CD4 T    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## B              0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD14+ Mono     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD8 T          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## NK             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Platelet       1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
## DC             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
##              142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157
## Naive CD4 T    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## B              0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD14+ Mono     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD8 T          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## NK             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Platelet       1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
## DC             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
##              158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173
## Naive CD4 T    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## B              0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD14+ Mono     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD8 T          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## NK             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Platelet       1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
## DC             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
##              174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189
## Naive CD4 T    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## B              0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD14+ Mono     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD8 T          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## NK             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Platelet       1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
## DC             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
##              190 191 192 193 194 195 196 197 198 199 200
## Naive CD4 T    0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono   0   0   0   0   0   0   0   0   0   0   0
## B              0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T   0   0   0   0   0   0   0   0   0   0   0
## CD14+ Mono     0   0   0   0   0   0   0   0   0   0   0
## CD8 T          0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED   0   0   0   0   0   0   0   0   0   0   0
## NK             0   0   0   0   0   0   0   0   0   0   0
## Platelet       1   1   1   1   1   1   1   1   1   1   1
## DC             0   0   0   0   0   0   0   0   0   0   0
## 
## $cellTypes
## [1] &quot;Platelet&quot; &quot;Platelet&quot; &quot;Platelet&quot;
</code></pre>
<h2 id="sec:max-change-attack">Max-change attack</h2>
<p>Modify the maximum number of genes inside a cell cluster without altering its classification.</p>
<p>The function <code>advMaxChange</code> runs a dichotomic search of gene subsets, given the cluster to attack and a type of modification, such that the classification does not change.</p>
<p>The <code>maxSplitSize</code> argument is the maximum size of dichotomic slices. Set to 1 to have better results, but it will take longer to compute.</p>
<pre><code class="language-r">genesMaxChange &lt;- advMaxChange(matPbmc, cellTypes, &quot;Memory CD4 T&quot;, MClassifier,
                    exclGenes = markers, advMethod = &quot;perc99&quot;)
</code></pre>
<pre><code class="language-r">length(genesMaxChange@values)
</code></pre>
<pre><code>## [1] 179
</code></pre>
<p>The function found 179 genes that you can modify with the <code>perc99</code> modification, and the cluster is still classified as <code>Memory CD4 T</code>.</p>
<p>Let’s run this attack and verify if it is successful.</p>
<p>First we modify the <code>matPbmc</code> matrix on the target cluster, on the genes previously determined.</p>
<pre><code class="language-r">matMaxAdver &lt;- advModifications(matPbmc, genesMaxChange@values, cellTypes, &quot;Memory CD4 T&quot;)
</code></pre>
<p>Then we verify that classification is still <code>Memory CD4 T</code>.</p>
<pre><code class="language-r">resClassif &lt;- MClassifier(matMaxAdver, cellTypes, &quot;Memory CD4 T&quot;)
</code></pre>
<pre><code class="language-r">resClassif
</code></pre>
<pre><code>## $prediction
## [1] &quot;Memory CD4 T&quot;
## 
## $odd
## [1] 1
## 
## $typePredictions
##              1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
## Naive CD4 T  0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## FCGR3A+ Mono 0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## B            0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Memory CD4 T 1 1 1 1 1 1 1 1 1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
## CD14+ Mono   0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## CD8 T        0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## UNDETERMINED 0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## NK           0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Platelet     0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## DC           0 0 0 0 0 0 0 0 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
##              26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
## Naive CD4 T   0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## FCGR3A+ Mono  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## B             0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Memory CD4 T  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
## CD14+ Mono    0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## CD8 T         0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## UNDETERMINED  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## NK            0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Platelet      0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## DC            0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
##              48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69
## Naive CD4 T   0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## FCGR3A+ Mono  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## B             0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Memory CD4 T  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
## CD14+ Mono    0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## CD8 T         0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## UNDETERMINED  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## NK            0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Platelet      0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## DC            0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
##              70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91
## Naive CD4 T   0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## FCGR3A+ Mono  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## B             0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Memory CD4 T  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
## CD14+ Mono    0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## CD8 T         0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## UNDETERMINED  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## NK            0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## Platelet      0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
## DC            0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
##              92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109
## Naive CD4 T   0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## B             0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T  1  1  1  1  1  1  1  1   1   1   1   1   1   1   1   1   1   1
## CD14+ Mono    0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## CD8 T         0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED  0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## NK            0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## Platelet      0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
## DC            0  0  0  0  0  0  0  0   0   0   0   0   0   0   0   0   0   0
##              110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
## Naive CD4 T    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## B              0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
## CD14+ Mono     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD8 T          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## NK             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Platelet       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## DC             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
##              126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141
## Naive CD4 T    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## B              0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
## CD14+ Mono     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD8 T          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## NK             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Platelet       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## DC             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
##              142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157
## Naive CD4 T    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## B              0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
## CD14+ Mono     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD8 T          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## NK             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Platelet       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## DC             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
##              158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173
## Naive CD4 T    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## B              0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
## CD14+ Mono     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD8 T          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## NK             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Platelet       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## DC             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
##              174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189
## Naive CD4 T    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## B              0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
## CD14+ Mono     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## CD8 T          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## NK             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## Platelet       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
## DC             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
##              190 191 192 193 194 195 196 197 198 199 200
## Naive CD4 T    0   0   0   0   0   0   0   0   0   0   0
## FCGR3A+ Mono   0   0   0   0   0   0   0   0   0   0   0
## B              0   0   0   0   0   0   0   0   0   0   0
## Memory CD4 T   1   1   1   1   1   1   1   1   1   1   1
## CD14+ Mono     0   0   0   0   0   0   0   0   0   0   0
## CD8 T          0   0   0   0   0   0   0   0   0   0   0
## UNDETERMINED   0   0   0   0   0   0   0   0   0   0   0
## NK             0   0   0   0   0   0   0   0   0   0   0
## Platelet       0   0   0   0   0   0   0   0   0   0   0
## DC             0   0   0   0   0   0   0   0   0   0   0
## 
## $cellTypes
##  [1] &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot;
##  [6] &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot;
## [11] &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot;
## [16] &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot;
## [21] &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot; &quot;Memory CD4 T&quot;
</code></pre>
<h2 id="sec:cgd-attack">CGD attack</h2>
<p>Apply the Cluster Gradient Descent (CGD) attack on the “Memory CD4 T” cells cluster with parameters alpha=epsilon=1</p>
<pre><code class="language-r">resCGD &lt;- advCGD(as.data.frame(matPbmc), cellTypes,
                &quot;Memory CD4 T&quot;, MClassifier, alpha=1, epsilon=1,
                genes=colnames(matPbmc)[ncol(matPbmc):1])
</code></pre>
<p>The algorithm test each gene and stops when the cluster prediction has switched
to another cell type prediction. Here it stops after having modified the CCR5 gene which make the classifier switch its prediction to Naive CD4 T.</p>
<pre><code class="language-r">tail(resCGD$byGeneSummary)
</code></pre>
<pre><code>##    order   gene nbModifiedCells oriCell_pourc targCell_pourc oriCell_odd
## 9      8   CD8A               0             1              0           1
## 10     9  MS4A1               0             1              0           1
## 11    10 S100A4               0             1              0           1
## 12    11    LYZ               0             1              0           1
## 13    12   CD14               0             1              0           1
## 14    13   CCR7              24             0              1         NaN
##    targCell_odd wholeClust_oriOdd wholeClust_targOdd
## 9             0               NaN                NaN
## 10            0               NaN                NaN
## 11            0               NaN                NaN
## 12            0               NaN                NaN
## 13            0               NaN                NaN
## 14            1               NaN                NaN
</code></pre>
<pre><code class="language-r">resCGD$oneRowSummary
</code></pre>
<pre><code>##  [1] &quot;Memory CD4 T&quot; &quot;24&quot;           &quot;Naive CD4 T&quot;  &quot;1&quot;            &quot;1&quot;           
##  [6] &quot;0&quot;            &quot;1&quot;            &quot;13&quot;           &quot;1&quot;            &quot;0&quot;           
## [11] &quot;0&quot;            &quot;1&quot;            &quot;NaN&quot;          &quot;NaN&quot;          &quot;NaN&quot;         
## [16] &quot;NaN&quot;
</code></pre>
<p>You can retrieve the modified matrix.</p>
<pre><code class="language-r">modifiedMatrix &lt;- resCGD$expr
</code></pre>
<p>And visualize the list of modified genes.</p>
<pre><code class="language-r">resCGD$modGenes
</code></pre>
<pre><code>## [1] &quot;CCR7&quot;
</code></pre>
<p>Check the new classification of the modified matrix.</p>
<pre><code class="language-r">MClassifier(modifiedMatrix, cellTypes, &quot;Memory CD4 T&quot;)$prediction
</code></pre>
<pre><code>## [1] &quot;Naive CD4 T&quot;
</code></pre>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.5.0 RC (2025-04-04 r88126)
## Platform: x86_64-pc-linux-gnu
## Running under: Ubuntu 24.04.2 LTS
## 
## Matrix products: default
## BLAS:   /home/biocbuild/bbs-3.21-bioc/R/lib/libRblas.so 
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.12.0  LAPACK version 3.12.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB              LC_COLLATE=C              
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## time zone: America/New_York
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] LoomExperiment_1.26.0       BiocIO_1.18.0              
##  [3] rhdf5_2.52.0                SingleCellExperiment_1.30.0
##  [5] SummarizedExperiment_1.38.0 Biobase_2.68.0             
##  [7] GenomicRanges_1.60.0        GenomeInfoDb_1.44.0        
##  [9] IRanges_2.42.0              MatrixGenerics_1.20.0      
## [11] matrixStats_1.5.0           S4Vectors_0.46.0           
## [13] BiocGenerics_0.54.0         generics_0.1.3             
## [15] adverSCarial_1.6.0         
## 
## loaded via a namespace (and not attached):
##  [1] Matrix_1.7-3            jsonlite_2.0.0          compiler_4.5.0         
##  [4] crayon_1.5.3            stringr_1.5.1           rhdf5filters_1.20.0    
##  [7] lattice_0.22-7          R6_2.6.1                XVector_0.48.0         
## [10] S4Arrays_1.8.0          knitr_1.50              DelayedArray_0.34.0    
## [13] h5mread_1.0.0           GenomeInfoDbData_1.2.14 rlang_1.1.6            
## [16] stringi_1.8.7           HDF5Array_1.36.0        xfun_0.52              
## [19] cli_3.6.4               SparseArray_1.8.0       magrittr_2.0.3         
## [22] Rhdf5lib_1.30.0         grid_4.5.0              lifecycle_1.0.4        
## [25] glue_1.8.0              evaluate_1.0.3          abind_1.4-8            
## [28] httr_1.4.7              tools_4.5.0             UCSC.utils_1.4.0
</code></pre>
</div>
</body>
</html>
